src/Jetpack_Usage.php

```diff
@@ -0,0 +1,182 @@
+				setcookie( 'tk_ai', 'expired', time() - 1000 );
+			}
+		}
+
+		$connection_manager = new Connection_Manager();
+		$wpcom_user_data = $connection_manager->get_connected_user_data( $user_id );
+		update_user_meta( $user_id, 'jetpack_tracks_wpcom_id', $wpcom_user_data['ID'] );
+
+		$this->tracking->record_user_event( 'wpa_user_linked', array() );
+	}
+
+	/**
+	 * Track that we've begun verifying the secrets.
+	 *
+	 * @access public
+	 *
+	 * @param string $action Type of secret (one of 'register', 'authorize', 'publicize').
+	 * @param \WP_User $user The user object.
+	 */
+	public function track_jetpack_verify_secrets_begin( $action, $user ) {
+		$this->tracking->record_user_event( "jpc_verify_{$action}_begin", array(), $user );
+	}
+
+	/**
+	 * Track that we've succeeded in verifying the secrets.
+	 *
+	 * @access public
+	 *
+	 * @param string $action Type of secret (one of 'register', 'authorize', 'publicize').
+	 * @param \WP_User $user The user object.
+	 */
+	public function track_jetpack_verify_secrets_success( $action, $user ) {
+		$this->tracking->record_user_event( "jpc_verify_{$action}_success", array(), $user );
+	}
+
+	/**
+	 * Track that we've failed verifying the secrets.
+	 *
+	 * @access public
+	 *
+	 * @param string $action Type of secret (one of 'register', 'authorize', 'publicize').
+	 * @param \WP_User $user The user object.
+	 * @param \WP_Error $error Error object.
+	 */
+	public function track_jetpack_verify_secrets_fail( $action, $user, $error ) {
+		$this->tracking->record_user_event(
+			"jpc_verify_{$action}_fail",
+			array(
+				'error_code'    => $error->get_error_code(),
+				'error_message' => $error->get_error_message(),
+			),
+			$user
+		);
+	}
+
+	/* Failed login attempts */
+	/**
+	 * Track a failed login attempt.
+	 *
+	 * @access public
+	 *
+	 * @param string $login Username or email address.
+	 */
+	public function track_failed_login_attempts( $login ) {
+		require_once JETPACK__PLUGIN_DIR . 'modules/protect/shared-functions.php';
+		$this->tracking->record_user_event(
+			'failed_login',
+			array(
+				'origin_ip' => jetpack_protect_get_ip(),
+				'login'     => $login,
+			)
+		);
+	}
+
+	/**
+	 * Track a connection failure at the registration step.
+	 *
+	 * @access public
+	 *
+	 * @param string|int $error      The error code.
+	 * @param \WP_Error  $registered The error object.
+	 */
+	function jetpack_connection_register_fail( $error, $registered ) {
+		$this->tracking->record_user_event( 'jpc_register_fail', array(
+			'error_code'    => $error,
+			'error_message' => $registered->get_error_message()
+		) );
+	}
+
+	function jetpack_connection_register_success( $from ) {
+		$this->tracking->record_user_event( 'jpc_register_success', array(
+			'from' => $from
+		) );
+	}
+
+	function jetpack_verify_api_authorization_request_error_double_encode() {
+		$this->tracking->record_user_event( 'error_double_encode' );
+	}
+
+	function jetpack_admin_ajax_tracks_callback() {
```

Comment:
/**
	 * Universal method for for all tracking events triggered via the JavaScript client.
	 *
	 * @access public
	 */

Reply:
1fbf56b04